File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
         ListNode temp = head;
        HashMap<ListNode, Integer> map = new HashMap<>();

        while (temp != null) {
            if (map.containsKey(temp)) {
                return true; // Cycle detected
            }

            map.put(temp, 1); // Mark as visited
            temp = temp.next;
        }

        return false; // No cycle found
        
    }
}

File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
          unordered_set<ListNode*> visited; // store addresses of visited nodes
        
        ListNode* current = head;
        while (current != nullptr) {
            // if we have already seen this node → cycle detected
            if (visited.find(current) != visited.end()) {
                return true;
            }
            visited.insert(current);  // mark node as visited
            current = current->next;
        }

        return false;
        
    }
};